transcom/mymove
# [MB-11454] PO9 - History and Audit Logging POC

## Summary

Proof of concept audit logging. The goal of PO9 is to create a history logging system like in Jira to show the progression of move changes. 

It will include things such as: who changed it, what it changed to, and when did it change. 

This is PoC to prove out that this is feasible to capture using DB triggers and an audit log table using JSON columns

## Things to highlight

### DB Changes

The audit history table will save the state of the old_data and the changed_data in a json format. This allows us to provide flexibility for the table data and also meet the requirements for Transcom. Read more about it in [this confluence page](https://dp3.atlassian.net/wiki/spaces/MT/pages/1634992192/Draft+PO9+Technical+Implementation#Proposed-Database-Design-#3:-One-Log-Table)

Here are the column descriptions for the `audit_history` table 

| column name       | description                                                        |
|-------------------|--------------------------------------------------------------------|
| id                | log id                                                             |
| schema_name       | name of the schema                                                 |
| table_name        | name of the table (i.e addresses, mto_shipments)                   |
| relid             |                                                                    |
| object_id         | id of the table entity (i.e addresses id column)                   |
| session_userid    | id of the user who made the change                                 |
| event_name        | the name of the API call (will be a FKey in actual implementation) |
| action_tstamp_tx  | Transaction start time                                             |
| action_tstamp_stm | Statement start time                                               |
| action_tstamp_clk | Trigger start time                                                 |
| transaction_id    | Identifier of transaction                                          |
| client_query      | The query executed                                                 |
| action            | Insert, update, or delete                                          |
| old_data          | State of the old data before the change                            |
| changed_data      | The change itself                                                  |
| statement_only    | True if audit event is from for each statement, false if not       |

Data gets inserted into this table via triggers. The trigger will evaluate whether it is a create, update, or delete and then create an audit record accordingly. Also, the update record is only created if there is a diff between the old record and the new one, so no empty `changed_data` records will be created for update. 

The `if_modified_func()` creates the new audit history record, it has separate `if` statements for `insert` `update` and `delete`

`add_audit_history_table` adds the triggers for these tables. We don't have to explicitly assign the triggers to the tables. 

It's hilarious: our custom parser in `pkg/migrate` can't handle the SQL in the migration file. If you paste in the migration manually, everything works

### API Handler Changes

The Application Context (`AuditableAppContextFromRequest`) now saves the session id of the user so that when transactions are made through the API, the DB can know who made the change. In addition, we allow the API to send the type of event is such as updateShipment to tell the DB the context of the change. 

The `AuditableAppContextFromRequest` makes use of 
`SET LOCAL audit.current_user_id = ?` and
`SET LOCAL audit.current_event_name = ?` to have the session user ID and  event_name available.

**For infrasec**, we would like them to review these specific changes and evaluate any security risks from adding this functionality to the API

## Steps to test and see API `moveHistory` response
1. Log into the office
2. Select a move from the queue and click on the  `History`  tab. On the `History` tab and you will  see the `JSON` out from the `moveHistory` API call.

<img width="865" alt="Screen Shot 2022-02-09 at 9 33 29 AM" src="https://user-images.githubusercontent.com/1359520/153234095-b6403922-704a-4ce5-9544-b9c8e9b76d7d.png">

3.  Fields to note in the `JSON` are the `changedValues`  field and the `oldValues` fields.  These  keys hold the old values before the update and the new values after the update.

## Review Screen mocks
* [history log usability miro board](https://miro.com/app/board/uXjVOQa6Hm4=/?moveToWidget=3458764518462448553&cot=14)
* [figma prototype from usability session](https://www.figma.com/proto/tGeEcWtsX9PjlPkmbhkJ4L/History-log-usability-test?page-id=0%3A1&node-id=2%3A3&scaling=scale-down-width)
